### DISCUSSION TITLE
How can I see the injected active and reactive power at a specific node?

### DISCUSSION THREAD

**Discussion by cqss11** [CURENT/andes#471] [Q&A] [answered]
How can I see the injected active and reactive power at a specific node? I couldn't find it in the ANDES user manual and I only see the bus.a or.v.
Thanks for your reply.

**Comment by jinningwang** [comment]
We don't have variables to store injected power of each node. The reason is to reduce unnecessary memory consumption.
 You can get the power injection with power flow equations with bus voltage and angle. Feel free to check the Example [working with data](https://docs.andes.app/en/latest/_examples/ex2.html#working-with-data).

**Reply by cqss11** [reply]
Thanks for your reply. And can I see the initial power of a certain node at a specific location in the code?

**Reply by jinningwang** [reply]
After calculating the injected power, you can get the entire time series data, where you can find the initial value.

**Comment by cuihantao** [comment]
Jinning,

Do we have a documentation mentioning that line injections are the Line.a1.e, Line.a2.e, Line.v1.e, and Line.v2.e fields?

For the original question, one can use a perturbation file to access these fields at the end of each step. Maybe you can help develop an example script.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by jinningwang** [reply]
I don't see such discussion in the existing documentation and examples. I'll do the example script and make a PR recently.

**Comment by jinningwang** [comment]
An example is attached here, [Power Injection.pdf](https://github.com/CURENT/andes/files/13062195/Power.Injection.pdf)

@cuihantao, would you kindly verify if I have correctly understood your idea?

**Reply by cuihantao** [reply]
Jinning,

This is clearly written. But why all the total P injections do not sum up to zero?

**Reply by jinningwang** [reply]
Updated one is attached, now the summed power is zero.

The reason is ``find_idx`` only returns first occurrence when multiple matches are there. For example, both Line_1 and Line_2 have "bus1" as "1", ``find_idx`` will only return Line_1.

**Reply by cuihantao** [reply]
Thanks a lot, Jinning.

I guess uses might also want to obtain these numbers at each step. They will then be able to obtain, for example, line injections following a disturbance. Can you also provide an example? You can use a pert file or come up with other methods.

**Comment by jinningwang** [comment, marked as answer]
Here, a further clarified example is attached, [Power Injection.pdf](https://github.com/CURENT/andes/files/13071739/Power.Injection.pdf)

The `pert.py` code is excerpted below:
```python
def pert(t, system):
    """
    Demonstration of perturbation function.

    This file should be applied to built-in IEEE 14-bus
    case "ieee14/ieee14_full.xlsx".

    In the given case, `Line` "Line_1" and "Line_2" are
    connected to `Bus` "1" through parameter `bus1`,
    and line injected active power is inspected through
    `Line.a1.e`.
    """
    Pinj_line = system.Line.get(src='a1', attr='e',
                                idx=['Line_1', 'Line_2'])
    print(f"t={t:.5f}, pinj_line={Pinj_line.sum():.5f}")
```

**Reply by cqss11** [reply]
Thanks a lot. I will check it!

**Reply by MuaYoi** [reply]
Hello, Mr Wang. I'm surprised why the injection power of all nodes is zero, which seems to lose its computational significance.

**Reply by jinningwang** [reply]
I believe given nodal power balance, all nodes have net zero power injection. Did you refer to line loss?

**Reply by MuaYoi** [reply]
Thanks for your response! There were some problems with my previous statement. Now let me put it another way, how to calculate the power of a bus flowing into a branch in time-domain simulation, that is, the power of the generator connected to the bus minus the power of the load connected to the bus.
And I don't understant the file pert.py, could you tell me where I can see the relevant documents?

**Reply by jinningwang** [reply]
1. The above discussion is about Line injected power, which I believe is what you expect.
2. https://docs.andes.app/en/latest/getting_started/formats/disturbance.html?highlight=pert#perturbation-file

**Reply by MuaYoi** [reply]
Thank you, Wang. I have read the above discussion again. I apologize for still having some areas that I don't understand. 
1. The pert.py pointed that line injected active power is inspected through `Line.a1.e`. What is Line.a1.e? The document says "e for equation value", I don't know what is it of the from bus's angle.
2. How can I calculate the injected reactive power?
3. To my knowledge, calculating the above physical quantities requires the node admittance matrix. Where can I obtain it?

Thank you very much, I wish you all the best!

**Reply by jinningwang** [reply]
You don't have to apologize to me or anyone  else here as long as the discussion follows the community [code of conduct](https://github.com/CURENT/andes/blob/master/CODE_OF_CONDUCT.md). The discussion boards aim to share knowledge about ANDES that might be helpful.

1. For ``ExtAlgeb``, the value of ``e`` represents the injected value. This is kind of tricky and more details can be found in the Development Documentation [Value and Equation Strings](https://docs.andes.app/en/latest/modeling/variables.html#value-and-equation-strings).
2. For a Line, injected active power is stored in ``a1`` and ``a2``, whereas reactive power is stored in ``v1`` and ``v2``.
3. The documentation of ``Line`` lists the used equations and they might answer you question.

**Reply by MuaYoi** [reply]
Thanks a lot, Wang. I code a pert.py to calculate what I want like this:
`def pert(t,system):
    Pinj_line = system.StaticLoad.get(src='bus',attr='e',idx=['PQ_1'])
    pass`
But how can I use the variable Pinj_line outside the function pert?
And another question is how I can get timing data for a specific device in `system.dae.ts.x` or `system.dae.ts.y`. It seems that I am ablet to get only one kind of data like `system.GENROU.Pe`, but actually I only want the Pe of GENROU_1.

**Reply by jinningwang** [reply]
1. Good question. To the best of my knowledge, `pert` function is not designed to have any return value and the function call `callpert` does not provide any return interface, as shown below. Then I guess there is not a direct way to feedback the pinj_line value outside the pert function.
https://github.com/CURENT/andes/blob/620373198bfcbfa8825020d3edc38ae195bea3fe/andes/routines/tds.py#L386
2. Actually example 2 in the documentation can answer this question, feel free to check it. https://docs.andes.app/en/latest/_examples/ex2.html#working-with-data

**Reply by MuaYoi** [reply]
Hi, Wang. I have read the section extracting variable data, in that example, the code `ss.GENROU.omega` returns an array `[4,5,6,7]`. I know that the array represents all GENROU's omega in system ss, but it seems that I can't work out which generator each element in the array corresponds to.

**Reply by jinningwang** [reply]
I believe in this section, https://docs.andes.app/en/latest/_examples/ex2.html#extracting-variable-data, desired variables number can be found, then the desired section of variables can be extracted.

**Reply by MuaYoi** [reply]
Hi Wang!
I am happy that I have found a way. If I want to find omega data of GENROU_1, I will use `ps.GENROU.get(attr='a',src='omega',idx='GENROU_1')` to get a index in ps.dae.ts.x, then I do what I want.
Thanks for your help and wish you a happy life!

**Comment by cuihantao** [comment]
@MuaYoi 

> But how can I use the variable Pinj_line outside the function pert?

It is unclear how "outside the function" means. Do you want to use it after the simulation is done, or do use the values at each step? You can provide at least some pseudo code for how you want to use it. 

>And another question is how I can get timing data for a specific device in system.dae.ts.x or system.dae.ts.y. It seems that I am ablet to get only one kind of data like system.GENROU.Pe, but actually I only want the Pe of GENROU_1.

The timestamps are the same for all the variables.

Like Jinning mentioned, https://docs.andes.app/en/latest/_examples/ex2.html#extracting-variable-data should give you some good ideas. In addition, the columns of `ss.dae.ts.y[:, ss.GENROU.v.Pe]` **correspond** to generators with indices `ss.GENTOU.idx.v`.

**Reply by MuaYoi** [reply]
Thanks for response Dr. Cui.
1. I want to use it after the simulation is done, just store `Pinj_line` of every step in an array, and I want to use the array to do something.
2. I have found a way. If I want to find omega data of GENROU_1, I will use `ps.GENROU.get(attr='a',src='omega',idx='GENROU_1')` to get a index in ps.dae.ts.x, then I do what I want.
Thanks for your help and wish you a happy life!

**Reply by jinningwang** [reply]
For 1, a quick way would be store the desired data in a CSV file, as the pert function leaves nothing after calling. The drawback is I/O with hard-disk at every time step can significantly slow down the simulation.
A more hacker way can be revise the pert function and corresponding calling function in `tds.py` to leave a interface parameter that can be stored and returned after simulation.

**Reply by MuaYoi** [reply]
Thanks a lot, maybe I will try it.

**Comment by cuihantao** [comment]
Another easier way is to use Jupyter so that you have an interactive session.

Next, in the pert function, duck tape the ss object by saving whatever you need as a field of ss. For example, ss.temp = data.

When done, you can access ss.temp retrieve the data. This is a commonly used hacking technique in Python.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by MuaYoi** [reply]
Thanks a lot. I will try it.
